xdsl/interpreter.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from io import StringIO
+from typing import Any, Callable, Generator, TypeVar
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import SSAValue, Operation
+
+_OperationInvT = TypeVar('_OperationInvT', bound=Operation)
+
+
+class IntepretationError(Exception):
+    pass
+
+
+@dataclass
+class Functions:
+    functions: dict[type[Operation],
```

Comment:
IMO, it is quite confusing that this is named `functions` while you also have functions in the toy dialect. Same will be true fro Func.Func at some point.

Not sure we can really change that. Maybe interpretation functions or something?

Reply:
I initially called it FunctionTable, maybe I should rename it back to that? It dispatches based on the type of the argument, as opposed to the callee, but sort of makes sense?

Reply:
- [ ] rename this